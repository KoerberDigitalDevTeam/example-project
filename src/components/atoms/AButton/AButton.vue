<template>
  <v-btn
    v-testId.btn
    v-bind="[$props, $attrs]"
    class="a-btn"
    rounded
    :color="getColor"
    :outlined="getOutlined"
    :text="getText"
    v-on="$listeners"
  >
    <slot />
  </v-btn>
</template>

<script>
export default {
  name: 'AButton',
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },

    primary: {
      type: Boolean,
      default: true,
    },

    secondary: {
      type: Boolean,
    },

    tertiary: {
      type: Boolean,
    },

    color: {
      type: String,
      default: undefined,
    },
  },

  computed: {
    primaryTheme() {
      return {
        color: 'primary',
        outlined: false,
        text: false,
      };
    },

    secondaryTheme() {
      return {
        color: 'dPrimary',
        outlined: true,
        text: false,
      };
    },

    tertiaryTheme() {
      return {
        color: 'dPrimary',
        outlined: false,
        text: true,
      };
    },

    currentTheme() {
      if (this.tertiary) {
        return this.tertiaryTheme;
      }
      if (this.secondary) {
        return this.secondaryTheme;
      }

      return this.primaryTheme;
    },

    getColor() {
      return this.color || this.currentTheme.color;
    },

    getOutlined() {
      return this.currentTheme.outlined;
    },

    getText() {
      return this.currentTheme.text;
    },
  },
};
</script>
<style lang="scss" scoped>
.a-btn {
  text-transform: none;

  ::v-deep .v-btn__content {
    font-family: Koerber, sans-serif;
    font-weight: 700;
    font-size: 13px;
  }
}
</style>
