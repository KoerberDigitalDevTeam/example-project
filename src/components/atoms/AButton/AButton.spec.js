import { shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import AButton from './AButton.vue';

describe('AButton', () => {
  function createWrapper(options = {}) {
    return shallowMount(AButton, {
      vuetify: new Vuetify(),
      ...options,
    });
  }

  describe('Template:', () => {
    it('Should render a AButton', () => {
      const wrapper = createWrapper();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should have a default slot', () => {
      const wrapper = createWrapper({
        slots: {
          default: 'default',
        },
      });

      expect(wrapper.text()).toContain('default');
    });
  });

  describe('Props Interface:', () => {
    it('Should have disabled prop required and must be a Boolean', () => {
      expect(AButton.props).toMatchObject({
        disabled: {
          type: Boolean,
          default: false,
        },
      });
    });

    it('Should have primary prop and must be a Boolean', () => {
      expect(AButton.props).toMatchObject({
        primary: {
          type: Boolean,
          default: true,
        },
      });
    });

    it('Should have secondary prop and must be a Boolean', () => {
      expect(AButton.props).toMatchObject({
        secondary: {
          type: Boolean,
        },
      });
    });

    it('Should have tertiary prop and must be a Boolean', () => {
      expect(AButton.props).toMatchObject({
        tertiary: {
          type: Boolean,
        },
      });
    });
    it('Should have color prop and must be a String', () => {
      expect(AButton.props).toMatchObject({
        color: {
          type: String,
          default: undefined,
        },
      });
    });
  });

  describe('Computed:', () => {
    const primaryTheme = {
      color: 'primary',
      outlined: false,
      text: false,
    };

    const secondaryTheme = {
      color: 'dPrimary',
      outlined: true,
      text: false,
    };

    const tertiaryTheme = {
      color: 'dPrimary',
      outlined: false,
      text: true,
    };

    describe('currentTheme', () => {
      it('When the `primary` prop is true then Should return the `primaryTheme`', () => {
        const localThis = {
          primary: true,
          primaryTheme,
        };
        const result = AButton.computed.currentTheme.call(localThis);
        expect(result).toBe(primaryTheme);
      });

      it('When the `secondary` prop is true then Should return the `secondaryTheme`', () => {
        const localThis = {
          primary: false,
          secondary: true,
          secondaryTheme,
        };
        const result = AButton.computed.currentTheme.call(localThis);
        expect(result).toBe(secondaryTheme);
      });

      it('When the `tertiary` prop is true then Should return the `tertiaryTheme`', () => {
        const localThis = {
          primary: false,
          tertiary: true,
          tertiaryTheme,
        };
        const result = AButton.computed.currentTheme.call(localThis);
        expect(result).toBe(tertiaryTheme);
      });
    });

    describe('getColor:', () => {
      it('When the `primary` prop is true then Should return the default color', () => {
        const localThis = {
          primary: true,
          currentTheme: primaryTheme,
        };
        const result = AButton.computed.getColor.call(localThis);
        expect(result).toBe('primary');
      });

      it('When the `primary` prop is false then Should return the alternative color', () => {
        const localThis = {
          primary: true,
          currentTheme: secondaryTheme,
        };
        const result = AButton.computed.getColor.call(localThis);
        expect(result).toBe('dPrimary');
      });

      it('When the `color` prop is valid then Should return the color value passed by prop', () => {
        const localThis = {
          color: 'test',
        };
        const result = AButton.computed.getColor.call(localThis);
        expect(result).toBe('test');
      });
    });

    describe('getOutlined:', () => {
      it('When the `secondary` prop is true then Should return true', () => {
        const localThis = {
          primary: true,
          currentTheme: secondaryTheme,
        };
        const result = AButton.computed.getOutlined.call(localThis);
        expect(result).toBe(true);
      });

      it('When the `secondary` prop is false then Should return false', () => {
        const localThis = {
          primary: true,
          currentTheme: primaryTheme,
        };
        const result = AButton.computed.getOutlined.call(localThis);
        expect(result).toBe(false);
      });
    });

    describe('getText:', () => {
      it('When the `tertiary` prop is true then Should return true', () => {
        const localThis = {
          primary: true,
          currentTheme: tertiaryTheme,
        };
        const result = AButton.computed.getText.call(localThis);
        expect(result).toBe(true);
      });

      it('When the `tertiary` prop is false then Should return false', () => {
        const localThis = {
          primary: true,
          currentTheme: primaryTheme,
        };
        const result = AButton.computed.getText.call(localThis);
        expect(result).toBe(false);
      });
    });
  });
});
