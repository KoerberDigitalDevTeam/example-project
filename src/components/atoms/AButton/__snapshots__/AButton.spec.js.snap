// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AButton Template: Should render a AButton 1`] = `<v-btn-stub color="primary" rounded="true" tag="button" activeclass="" type="button" primary="true" class="a-btn" data-testid="btn"></v-btn-stub>`;
